src/main/scala/com/fulcrumgenomics/bam/FindSwitchbackReads.scala

```diff
@@ -0,0 +1,398 @@
+    |`--metrics=s1.switchback` the following files are produced:
+    |
+    |1. `s1.switchback.summary.txt`: A table of summary metrics describing the number of reads, switchbacks, etc.
+    |2. `s1.switchback.lengths.txt`: A table of the distribution of observed switchback lengths in read-based switchbacks.
+    |3. `s1.switchback.offsets.txt`: A table of the distribution of observed offsets in read-based switchbacks.
+    |4. `s1.switchback.gaps.txt`: A table of the distribution of gap lengths in tampl
+    |5. `s1.switchback.plots.pdf`: A PDF containing plots of the distributions from 2-4.
+    |
+    |Note: because this tool accesses the reference genome in a random manner it pre-loads the entire reference fasta
+    |into memory.  As a result the tool is best run with `-Xmx8g` to give it sufficient memory.
+  """)
+class FindSwitchbackReads
+( @arg(flag='i', doc="Input BAM file.") val input: PathToBam = Io.StdIn,
+  @arg(flag='o', doc="Output BAM file.") val output: PathToBam = Io.StdOut,
+  @arg(flag='m', doc="Metrics output file.") val metrics: Option[PathPrefix] = None,
+  @arg(flag='s', doc="Output sort order.") val sortOrder: Option[SamOrder] = None,
+  @arg(flag='r', doc="Reference genome fasta file.") val ref: PathToFasta,
+  @arg(flag='O', doc="Maximum offset between end the two segments of the read on the reference. Set to 0 to disable read-based checks.") val maxOffset: Int = 35,
+  @arg(flag='g', doc="Maximum gap between R1 and R2 of tandem reads to call a template a switchback. Set to 0 to disable tandem-based checks.") val maxGap: Int = 500,
+  @arg(flag='l', doc="Minimum match length of the switched back segment.") val minLength: Int = 6,
+  @arg(flag='e', doc="Maximum error rate of switchback match to genome.") val maxErrorRate: Double = 0.1,
+  @arg(flag='d', doc="IF true, do NOT unmap reads from switchback templates.") val dontUnmap: Boolean = false
+) extends FgBioTool with LazyLogging {
+  import FindSwitchbackReads._
+  Io.assertReadable(input)
+  Io.assertReadable(ref)
+
+  override def execute(): Unit = {
+    logger.info("Loading reference sequences.")
+    val refs = ReferenceSequenceIterator(ref, stripComments=true).map(r => r.getName -> r).toMap
+
+    val progress = ProgressLogger(logger)
+    var total   = 0L
+    var aligned = 0L
+    val switchbackGaps    = new NumericCounter[Int]()
+    val switchbackLengths = new NumericCounter[Int]()
+    val switchbackOffsets = new NumericCounter[Int]()
+
+    val in = SamSource(input)
+    val iterator = Bams.templateIterator(in)
+
+    val outHeader = in.header.clone()
+    sortOrder match {
+      case Some(so) => so.applyTo(outHeader)
+      case None     =>
+        if (in.header.getSortOrder != SortOrder.queryname && in.header.getGroupOrder != GroupOrder.query) {
+          SamOrder.Queryname.applyTo(outHeader)
+        }
+    }
+    sortOrder.foreach(_.applyTo(outHeader))
+    val out = SamWriter(output, header=outHeader, sort=sortOrder)
```

Comment:
Just my thought on style and scoping, but I'd prefer this:
```
val out = {
    val outHeader = in.header.clone()
    sortOrder match {
      case Some(so) => so.applyTo(outHeader)
      case None     =>
        if (in.header.getSortOrder != SortOrder.queryname && in.header.getGroupOrder != GroupOrder.query) {
          SamOrder.Queryname.applyTo(outHeader)
        }
    }
    sortOrder.foreach(_.applyTo(outHeader))
    SamWriter(output, header=outHeader, sort=sortOrder)
}
```
For me, it keeps objects tightly scoped.

Reply:
If you want to go all the way, which you don't, you could do:
```
val out = {
    val outHeader = {
        val header = in.header.clone()
        sortOrder match {
          case Some(so) => so.applyTo(header)
          case None     =>
            if (in.header.getSortOrder != SortOrder.queryname && in.header.getGroupOrder != GroupOrder.query) {
              SamOrder.Queryname.applyTo(header)
            }
        }
        sortOrder.foreach(_.applyTo(header))
    }
    SamWriter(output, header=outHeader, sort=sortOrder)
}
```